miniapps/shifted/marking.cpp

```diff
@@ -0,0 +1,333 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "marking.hpp"
+
+namespace mfem
+{
+
+void ShiftedFaceMarker::MarkElements(Array<int> &elem_marker) const
+{
+   elem_marker.SetSize(pmesh.GetNE() + pmesh.GetNSharedFaces());
+   elem_marker = SBElementType::INSIDE;
+
+   IntegrationRules IntRulesLo(0, Quadrature1D::GaussLobatto);
+
+   Vector vals;
+   // Check elements on the current MPI rank
+   for (int i = 0; i < pmesh.GetNE(); i++)
+   {
+      ElementTransformation *Tr = pmesh.GetElementTransformation(i);
+      const IntegrationRule &ir =
+         IntRulesLo.Get(pmesh.GetElementBaseGeometry(i), 4*Tr->OrderJ());
+      ls_func.GetValues(i, ir, vals);
+
+      int count = 0;
+      for (int j = 0; j < ir.GetNPoints(); j++)
+      {
+         if (vals(j) <= 0.) { count++; }
```

Comment:
I wonder if a strict condition like this will work in general, or if a tolerance should be used (`vals(j) < tol`). If there is round-off or numerical error, points on the boundary of an element may not be well-defined by this condition without a tolerance (if the integration rule has boundary points or points very close to the boundary). I am not familiar with `HeatDistanceSolver`, but could that have numerical error that may require using a tolerance? If a tolerance is necessary, defining it in general for any problem may be challenging.